patches/tModLoader/Terraria/ModLoader/ChestLoot.cs

```diff
@@ -0,0 +1,321 @@
+			bool hasInNormal = ind2.item != defaultInstance;
+			if (hasInNormal)
+				return ind2;
+
+			return putCandidates2.ElementAtOrDefault(index);
+		}
+	}
+
+	public ChestLoot() {
+		items = new();
+		putCandidates = new();
+		putCandidates2 = new();
+	}
+
+	private void AddCandidates(IList<Entry> entries) {
+		IReadOnlyList<(int nextTo, bool after)> candidates = putCandidates;
+		candidates = candidates.Reverse().ToList();
+
+		List<(int nextTo, bool after)> unsuccessedCandidates = new();
+
+		for (int i = 0; i < 2; i++) { // run twice, cause mods might want to add to chained items
+			var a = candidates;
+			if (i == 1)
+				a = unsuccessedCandidates;
+
+			foreach (var (nextTo, after) in a) {
+				int index = items.FindIndex(x => x.item.type.Equals(nextTo));
+				if (index != -1) {
+					entries.Insert(index + after.ToInt(), putCandidates2[index]);
+				}
+				else {
+					unsuccessedCandidates.Add((nextTo, after));
+				}
+			}
+		}
+	}
+
+	public void AddRange(params Entry[] entries) {
+		foreach (var e in entries) {
+			Add(e);
+		}
+	}
+
+	public void Add(Entry entry) {
+		items.Add(entry);
+	}
+
+	public bool Add(int item, params ICondition[] condition) {
+		return Add(ContentSamples.ItemsByType[item], condition);
+	}
+
+	public bool Add(Item item, params ICondition[] condition) {
+		Add(new Entry(item, condition));
+		return true;
+	}
+
+	private bool PutAt(int destination, Item item, bool after, params ICondition[] condition) {
+		putCandidates.Add(new(destination, after));
+		putCandidates2.Add(new(item, condition));
+		return true;
+	}
+
+	private bool PutAt(int destination, int item, bool after, params ICondition[] condition) {
+		return PutAt(destination, ContentSamples.ItemsByType[item], after, condition);
+	}
+
+	public bool InsertBefore(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertBefore(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertAfter(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
+
+	public bool InsertAfter(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
+
+	public bool Hide(int item) {
+		Entry entry = this[item];
+		entry.Hidden = true;
+		return true;
+	}
+
+	public Item[] Build(bool lastSlotEmpty = true) {
+		return Build(out _, lastSlotEmpty);
+	}
+
+	public Item[] Build(out int slots, bool lastSlotEmpty = true) {
+		List<Item> array = new();
+		List<Entry> oldEntries = items; // incase current instance still gets used after building for some reason.
+
+		AddCandidates(oldEntries);
+		foreach (Entry group in oldEntries) {
+			group.TryAdd(array);
+		}
+
+		slots = array.Count;
+		if (array.Count < 40) {
+			array.AddRange(Enumerable.Repeat(emptyInstance, 40 - array.Count));
+		}
```

Comment:
Add additional check for array.Count > 40, and if so, output warning to log and chat that all items could not fit in the shop